src/ynEIGEN/WithdrawalsProcessor.sol

```diff
@@ -0,0 +1,432 @@
+                        ? _availableToWithdraw
+                        : _pendingWithdrawalRequestsInShares;
+                    _shares[i] = _toWithdraw;
+                    _pendingWithdrawalRequestsInShares -= _toWithdraw;
+                }
+            }
+
+            // second pass: withdraw evenly from all nodes if there is still more to withdraw
+            if (_pendingWithdrawalRequestsInShares > minPendingWithdrawalRequestAmount) {
+                // NOTE: here we compare shares to unit... nbd?
+                uint256 _equalWithdrawal = _pendingWithdrawalRequestsInShares / _nodesLength;
+                for (uint256 i = 0; i < _nodesLength; ++i) {
+                    _shares[i] = _equalWithdrawal + MIN_DELTA > _nodesShares[i] ? _nodesShares[i] : _equalWithdrawal;
+                }
+            }
+        }
+    }
+
+    //
+    // mutative functions
+    //
+
+    /// @notice Queues withdrawals
+    /// @dev Reverts if the total pending withdrawal requests are below the minimum threshold
+    /// @dev Saves the queued withdrawals together in a batch, to be completed in the next step (`completeQueuedWithdrawals`)
+    /// @dev Before calling this function, call `getQueueWithdrawalsArgs()` to get the arguments
+    /// @param _asset The asset to withdraw
+    /// @param _nodes The list of nodes to withdraw from
+    /// @param _amounts The share amounts to withdraw from each node
+    /// @return True if all pending withdrawal requests were queued, false otherwise
+    function queueWithdrawals(
+        IERC20 _asset,
+        ITokenStakingNode[] memory _nodes,
+        uint256[] memory _amounts
+    ) external onlyRole(KEEPER_ROLE) returns (bool) {
+        uint256 _nodesLength = _nodes.length;
+        if (_nodesLength != _amounts.length) revert InvalidInput();
+
+        uint256 _pendingWithdrawalRequests = getPendingWithdrawalRequests(); // NOTE: reverts if too low
+        uint256 _toBeQueued = _pendingWithdrawalRequests;
+
+        IStrategy _strategy = ynStrategyManager.strategies(_asset);
+        if (_strategy == IStrategy(address(0))) revert InvalidInput();
+
+        uint256 _queuedId = _ids.queued;
+        uint256 _pendingWithdrawalRequestsInShares = _unitToShares(_pendingWithdrawalRequests, _asset, _strategy);
+        for (uint256 j = 0; j < _nodesLength; ++j) {
+            uint256 _toWithdraw = _amounts[j];
+            if (_toWithdraw > 0) {
+                _toWithdraw > _pendingWithdrawalRequestsInShares
+                    ? _pendingWithdrawalRequestsInShares = 0
+                    : _pendingWithdrawalRequestsInShares -= _toWithdraw;
+
+                address _node = address(_nodes[j]);
+                _queuedWithdrawals[_queuedId++] = QueuedWithdrawal(
+                    _node,
+                    address(_strategy),
+                    delegationManager.cumulativeWithdrawalsQueued(_node), // nonce
+                    _toWithdraw,
+                    withdrawalQueueManager._tokenIdCounter(),
```

Comment:
to verify my understanding of this

It's safe to note that all withdrawals are completed up to _tokenIdCounter because 

`queueWithdrawals` is called with an input that solves for all the current pending unfulfilled amount in one go, so you can mark them all as completed

is that correct?

Reply:
good point - not in a go, but the assumption here is that `queueWithdrawals` is called continuously until all requests are satisfied